src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1245 @@
+    BitSet const& nodes, BitSet const& disj) const
+{
+    mPotentialSplit.first.clear();
+    mPotentialSplit.second.clear();
+    CLOG(ERROR, "SCP") << "Found potential disjoint quorums";
+    CLOG(DEBUG, "SCP") << "Quorum A ids: " << nodes;
+    CLOG(DEBUG, "SCP") << "Quorum B ids: " << disj;
+    for (size_t i = 0; nodes.next_set(i); ++i)
+    {
+        CLOG(ERROR, "SCP") << "Quorum A: " << nodeName(i);
+        mPotentialSplit.first.emplace_back(mBitNumPubKeys.at(i));
+    }
+    CLOG(ERROR, "SCP") << "---";
+    for (size_t i = 0; disj.next_set(i); ++i)
+    {
+        CLOG(ERROR, "SCP") << "Quorum B: " << nodeName(i);
+        mPotentialSplit.second.emplace_back(mBitNumPubKeys.at(i));
+    }
+}
+
+bool
+QuorumIntersectionCheckerImpl::hasDisjointQuorum(BitSet const& nodes) const
+{
+    BitSet disj = contractToMaximalQuorum(mMaxSCC - nodes);
+    if (disj)
+    {
+        CLOG(TRACE, "SCP") << "found quorum  = " << nodes;
+        CLOG(TRACE, "SCP") << "disjoint with = " << disj;
+        noteFoundDisjointQuorums(nodes, disj);
+    }
+    else
+    {
+        if (mLogTrace)
+        {
+            CLOG(TRACE, "SCP")
+                << "no quorum in complement  = " << (mMaxSCC - nodes);
+        }
+    }
+    return disj;
+}
+
+void
+decrementIfNonZero(size_t& sz)
+{
+    if (sz > 0)
+    {
+        --sz;
+    }
+}
+
+QBitSet
+QuorumIntersectionCheckerImpl::convertSCPQuorumSet(SCPQuorumSet const& sqs)
+{
+    size_t threshold = sqs.threshold;
+    BitSet nodeBits(mPubKeyBitNums.size());
+    for (auto const& v : sqs.validators)
+    {
+        auto i = mPubKeyBitNums.find(v);
+        if (i == mPubKeyBitNums.end())
+        {
+            // This node 'v' is one we do not have a qset for. We simulate 'v'
+            // being of unknown state by both not-mentioning 'v' in the qset
+            // we're building (that depends on 'v'), and by reducing the
+            // threshold of the qset we're building by one: it'll pass _or_ fail
+            // with one-less bit of evidence overall.
+            decrementIfNonZero(threshold);
+        }
+        else
+        {
+            auto i = mPubKeyBitNums.find(v);
+            assert(i != mPubKeyBitNums.end());
+            nodeBits.set(i->second);
+        }
+    }
+    QGraph inner;
+    inner.reserve(sqs.innerSets.size());
+    for (auto const& i : sqs.innerSets)
+    {
+        auto qbInner = convertSCPQuorumSet(i);
+        if (qbInner.empty())
+        {
+            // Similarly if an innerSet has nothing in it (which happens when it
+            // recursively references only unknown nodes), we treat it as though
+            // it's not present either, eliding it from the converted inner
+            // QGraph and decrementing our threshold further.
+            decrementIfNonZero(threshold);
+        }
+        else
+        {
+            inner.emplace_back(qbInner);
+        }
+    }
+    return QBitSet(threshold, nodeBits, inner);
```

Comment:
I *think* that this is not exactly right and we need to decide what we want to do.

First, note that we are *excluding* nodes that don't have a quorum set from the "quorum intersection" criteria: reason for this is that, they could be misconfigured and only point to themselves for example. So what we're instead looking at is their potential to influence others as byzantine nodes which is a bit pessimistic, but we can't do a lot better than this.

Now, note that if `threshold == 0` it means that that node reached consensus based on some collection of other nodes.

For the "first generation nodes" that reached consensus just with nodes that don't have a quorum set, it means that those nodes will externalize (potentially different) values dictated by those "missing" nodes.

We need to decide what to do with them at this point:
* one way to look at it is, for one of those "first gen nodes", its quorum set should be replaced by a quorum set that only contains itself (it got convinced of badness but is consistent about it). If we have more than one such node, they may externalize different values (because the other guys are byzantine), so we have a split. We can ignore that as well if we want: we can replace the quorum set of the second node to just point to the first node (forcing those "first gen nodes" to pick the same value).
* another way, simpler (at first and maybe too pessimistic), would be to consider that such "first gen node" should now be also removed from the set of nodes we're looking at in the context of quorum intersection, and just do as if they didn't have a quorum set either (so a cascading byzantine failure type of situation). If we do this, the question that we need to answer is how far can we go like this? Like if we remove a large part of the network, why are we claiming quorum intersection?

Reply:
(Resolved offline)